### ISSUE TITLE
Rename parameters to make error messages clearer

### ISSUE THREAD

**Issue by pdcastro** [balena-io/balena-cli#1758]
Feedback from the balena team itself (most recently from @OdysLam but also from others) is that error messages regarding missing/incorrect command-line arguments/parameters are unclear in some cases.

Parameter names should be review to see if there are any more descriptive or more easily understood name choices.

**Comment by srlowe** [comment]
`in the command line` sounds a bit wordy to me.  It's this redundant?

**Comment by srlowe** [comment]
I think we should be careful when changing parameter names.  E.g. I think changing `balena push <applicationOrDevice>` to `<UUID>` would be incorrect, since it can accept an application name, an IP, or a .local domain name (no UUIDs AFAIK).

**Comment by OdysLam** [comment]
Maybe:
$ balena push
Proper ussage: balena push <argument>, type balena push --help for detailed instructions

Something along those lines, so that I can 1)easily see the right format of the command 2) know that I can invoke a specific command for a detailed view of the command's instructions.

**Comment by srlowe** [comment]
@OdysLam I think that approach would not be possible across all commands, since there are often multiple valid usages of one command.

**Comment by pdcastro** [comment]
@OdysLam, the complication is that the error message is not produced by "our code" directly, it is produced the CLI framework parser: either [Capitano](https://github.com/balena-io/capitano/) or [oclif](https://oclif.io/), depending on the command. It is a "standardized" message, currently in the format:
```
[Missing] [what]
```
Where [Missing] is hardcoded in the framework, and [what] is the variable name in "our code". We can patch the framework, and what I meant by "compromise" is to make the patch itself standardized and limited:
* By renaming variables in "our code", for example "UUID" instead of "applicationOrDevice"
* By patching the standardized message components to modify the prefix and add a suffix:
```
[Missing] [what] [where]
```
e.g.
```
[Missing] ['UUID' parameter] [in the command line]
```

Having said that, when we look at the frameworks' source code we will give it some thought at the extend to which we can customise them, so your suggestion is welcome. 👍

**Comment by srlowe** [comment]
@OdysLam @pdcastro  A compromise that might work in these missing parameter cases, could be to reuse the help usage examples.  E.g. Something like this:
```
$ balena push
Missing parameter 'applicationOrDevice'

Examples:

	$ balena push myApp
	$ balena push myApp --source <source directory>
	$ balena push myApp -s <source directory>

	$ balena push 10.0.0.1
	$ balena push 10.0.0.1 --source <source directory>
	$ balena push 10.0.0.1 --service my-service
	$ balena push 10.0.0.1 --env MY_ENV_VAR=value --env my-service:SERVICE_VAR=value
	$ balena push 10.0.0.1 --nolive

	$ balena push 23c73a1.local --system
	$ balena push 23c73a1.local --system --service my-service

type balena push --help for detailed instructions
```

**Comment by OdysLam** [comment]
I am inclined to say that it's cleaner to simply say what parameter misses or what argument is not recognizable and leave everything else for the --help.

Giving examples without explanation is moot since I will probably hit help either way.

**Comment by pdcastro** [comment]
> `in the command line` sounds a bit wordy to me. Isn't this redundant?

* _"Missing 'device'"_ is too short, could be confusing.
* _"Missing 'device' parameter"_ or _"Missing parameter 'device'"_ are certainly better, then the question is whether a CLI newbie would readily understand that 'parameter' means a command-line parameter.
* _"Missing 'device' parameter in the command line"_ may be somewhat redundant, but if it saves a newbie contacting balena's support, I don't mind being redundant!

**Comment by pdcastro** [comment]
> changing balena push <applicationOrDevice> to <UUID> would be incorrect, since it can accept an application name [...]

Indeed, UUID doesn't make any sense for `balena push`, I must have had a different command in mind. The principle is to find a "neutral" word that is easier to read / understand than "applicationOrDevice".

**Comment by pdcastro** [comment]
> A compromise that might work in these missing parameter cases, could be to reuse the help usage examples

This would be doing the opposite of what is requested in #1757 [Shorten error messages](https://github.com/balena-io/balena-cli/issues/1757) ... And we'd have to review the examples to make sure they cover all command line variations.

**Comment by OdysLam** [comment]
From what I understand, we can't custom define an error message for each command, like:
`YOu are missing the application name or the device ip or local name in case of local-mode`. Am I right?

I think a good generic would be: Missing `field` and then advise user to look into balena --help, then remove everything else.

**Comment by pdcastro** [comment]
Given the work in converting commands from Capitano to oclif (#1768), if there is any framework modification, let's limit it to oclif only. The benefit will reach all commands soon enough.

**Comment by srlowe** [comment]
(copied from FD)
BTW - this is this kind of thing that oclif currently produces:
```
Missing 1 required arg:
name  environment or config variable name
See more help with --help
```
What changes would you both want to that?

I think It's not bad, but:
 - the count seems a bit technical
 - `arg` could be expanded to `argument`
 - The param name isn't displayed very clearly - it kind of merges in with the param description.

**Comment by OdysLam** [comment]
Just to be sure that I understand it correctly, @srlowe can you give me an example with `balena push` ?

**Comment by srlowe** [comment]
`push` would currently look like this in oclif (I had to make up parameter description, since that concept doesn't exist in capitano).
```
$ balena push

Missing 1 required arg:
applicationOrDevice  application name, device IP, or device .local hostname
See more help with --help
```

**Comment by OdysLam** [comment]
I can't understand the connection between `applicationOrDevice` and the rest of the line. Can we either remove the `applicationOrDevice` or show that the rest of the line describes that argument?
@srlowe

**Comment by srlowe** [comment]
`applicationOrDevice` is the parameter name.  The rest of the line is the description of that parameter.

**Comment by OdysLam** [comment]
As a developer, I can deduce as such (although it's not very clear), but I explained you my view as a user who can have all sorts of backgrounds. @srlowe

**Comment by srlowe** [comment]
@pdcastro I suggest we close this issue as it's a bit of a mess (e.g it's covering things that have been resolved already). I think if there are issues with the naming of params for a particular command, then we should open individual issues, otherwise I think this issue will just sit here forever.
